### PR TITLE
Remove single_precision_histogram

### PR DESCRIPTION
**Pull request by RAMitchell**
Removing the `single_precision_histogram` due to major observed accuracy issues. In particular many tests can fail due to numerical error accumulating in gain calculation, resulting in leaf nodes being opened that contain no training data at all.

In this PR I have removed a now redundant class in gpu_hist and slightly refactored the TreeUpdater base class to always require a global context for construction.

### CODE REVIEW

**Comment by trivialfis** [comment]
Close https://github.com/dmlc/xgboost/issues/7728 .

File: src/tree/updater_gpu_hist.cu [hunk 1]
```diff
@@ -532,6 +529,13 @@ struct GPUHistMakerDevice {
 
   void ApplySplit(const GPUExpandEntry& candidate, RegTree* p_tree) {
     RegTree& tree = *p_tree;
+
+    // Sanity check - have we created a leaf with no training instances?
+    if (!rabit::IsDistributed() && row_partitioner) {
```
Comment (trivialfis) [inline src/tree/updater_gpu_hist.cu:534]: Any potential solution if this check occurs?

Comment (trivialfis) [reply src/tree/updater_gpu_hist.cu:534]: My mistake.  I ran into issues with empty nodes in partitioner `UpdatePosition`, where `left_count` can be 0 due to floating-point errors.  In this `ApplySplit` function, if we check on the left/right child instead of the parent (candidate), then we might run into trouble.

File: tests/python-gpu/test_gpu_updaters.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 import gc
 import pytest
 import xgboost as xgb
-from hypothesis import given, strategies, assume, settings, note
+from hypothesis import given, strategies, assume, settings, note, reproduce_failure
```
Comment (trivialfis) [inline tests/python-gpu/test_gpu_updaters.py:6]: I think this is forgotten.

**Review by trivialfis — COMMENTED** [review]
Apologies for the conflict.  Do you need help with rebasing?

**Review by trivialfis — APPROVED** [review]
